proposals/0000-splice-imports.rst

```diff
@@ -0,0 +1,319 @@
+   number of modules have to be cautiously compiled to bytecode.
+4. By using splice imports we can separate the dependencies into those only needed at build-time (1) and
+   those only needed at runtime (2). We can then link only against those packages needed at runtime.
+5. Currently, when cross-compiling, in modules that use ``TemplateHaskell``, all
+   imported modules must be compiled for both host and target.
+   By distinguishing imported modules used at compile time(i.e. not used at
+   runtime), we can require only those modules to be compiled for the host.
+   Similarly, by distinguishing imported modules used at runtime (i.e. not used at
+   compile time), we can require only those modules to be compiled for the
+   target. It can be very hard or impossible to make some packages available on
+   some cross-compile target platforms, so this change would significantly
+   improve the applicability of ``TemplateHaskell`` in these scenarios.
+
+
+Definitions
+-----------
+
+level
+  Each expression exists at a level. The level is increased by 1 when
+  inside a quote and decreased by 1 inside a splice. Therefore the level of
+  an expression can be calculated as the number of quotes surrounding an expression
+  subtract the number of splices. For example::
+
+    -- foo is at level 0
+    foo = $(let
+      -- bar is at level -1
+      bar = $(let
+        -- baz is at level -2
+        baz = [|
+        -- qux is at level -1
+          qux = [|
+            -- quux is at level 0
+            quux = [|
+              quuz is at level 1
+              quuz = 0
+            |]
+          |]
+        |] in baz
+      ) in bar
+    )
+
+
+top-level splice
+  A splice, where the body is at a negative level, an unadorned
+  declaration splice or a quasiquoter.
+
+
+home module
+  A module from the package that is currently being compiled.
+
+
+Proposed Change
+---------------
+
+The ``splice`` modifier indicates that a module's imports are available at compile time.
+The absence of the ``splice`` modifier indicates that a module's imports are available
+at runtime.
+
+A ``splice`` imported module allows identifiers to be used at compile time. Compile time
+evaluation is indicated by a top-level splice. Therefore the ``splice`` modifier
+affects name resolution depending on whether we are inside a top-level splice or not.
+
+1. If a module is only available at compile time then the imports are only available in top-level splices.
+2. If a module is only available at runtime then the imports are not available in top-level splices.
+3. If a module is available at both runtime and compile time then the imports are available everywhere.
+
+The new language extension ``ExplicitSpliceImports`` adds a
+new import modifier to the import syntax. An import is marked as a "splice"
+import when it is prefixed with ``splice``::
+
+  {-# LANGUAGE ExplicitSpliceImports #-}
+  {-# LANGUAGE TemplateHaskell #-}
+  module Main where
+
+  -- (1)
+  import splice B
```

Comment:
Bikeshedding: this reads as if you’re importing a splice called `B`. I know splices can’t even have names, but that’s what I see when I read this line. Have other pseudo-keywords been considered? Racket seems to be used `for-syntax`, maybe we want `for-splices` or something?

Since `$` is used for splices, perhaps `import $B` could work.